### PR TITLE
Prevent dirty service object leaking between reconciles

### LINKED ISSUE #59: Service Controller can call provider EnsureLoadBalancer with a dirty Service object
I observed this with cloud-provider-openstack, but I believe this is a framework bug not a provider bug. Specifically we had a permissions issue which caused this Patch to always fail: https://github.com/kubernetes/cloud-provider-openstack/blob/21ce6dd984e5599ba3c411c960acdbcf9b5dbf03/pkg/openstack/loadbalancer.go#L1783

However, completely deterministically this succeeds the second time it is called. There is no way for this reconcile to succeed without [setting an annotation](https://github.com/kubernetes/cloud-provider-openstack/blob/21ce6dd984e5599ba3c411c960acdbcf9b5dbf03/pkg/openstack/loadbalancer.go#L1907), but after the successful reconciliation we see that the annotation is not set. The only explanation I could muster for this is that we were called with the dirty object from the first reconciliation. Consequently the Patcher is based on a dirty object which already has the annotation set, even though we never succeeded to write that to the API, and therefore the Patcher believes second time round that it doesn't need to write the annotation and we erroneously succeed without having written the required annotation.

The obvious suspect was the [cache used by the service controller](https://github.com/kubernetes/kubernetes/blob/349900472a38a29fd6d85f7e4880d4f3d72ad6ee/staging/src/k8s.io/cloud-provider/controllers/service/controller.go#L334-L335), however for this specific call path we are always passing in a Service object which came from the informer.

...The informer code is hard to follow...

However, I believe our Service object is ultimately coming from a [ThreadSafeStore](https://github.com/kubernetes/kubernetes/blob/cbbb5f70a47644f9830073d9d0329bf247a328a1/staging/src/k8s.io/client-go/tools/cache/thread_safe_store.go#L63-L72). 2 things to note about this class:

* [Its documentation says not to modify anything returned by it](https://github.com/kubernetes/kubernetes/blob/cbbb5f70a47644f9830073d9d0329bf247a328a1/staging/src/k8s.io/client-go/tools/cache/thread_safe_store.go#L31-L33)
* All its accessors return a shallow copy of the cached item

I believe the second thing is the immediate cause of the issue which we hit in cloud-provider-openstack. Annotations, being a slice, are not copied. Therefore the annotations in the second invocation are the same as the first invocation.

Given that:

* the documentation says not to modify these items
* AFAIK it's not a bug for a reconciler to add an annotation to the object

I believe we need to deep copy the object returned by the informer before passing it to the sync loop.

### PR DESCRIPTION
**Pull request by mdbooth**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
We were passing a Service object returned from the Service informer directly to the Service controller loop. This loop is expected to modify the Service object during execution, but [this is not safe](https://github.com/kubernetes/kubernetes/blob/f0791b50143856177878e21bb44beb5e3e36cc78/staging/src/k8s.io/client-go/tools/cache/thread_safe_store.go#L26-L40). Specifically any modification to the Service object by the controller will pollute the Service informer's cache and result in a potentially dirty object being passed to subsequent invocation of the controller loop.

We fix this issue by passing a copy of the object from the informer instead.

I have described in the comments how this bug affected cloud-provider-openstack, but I suspect it affects multiple cloud providers.

#### Which issue(s) this PR fixes:
Fixes https://github.com/kubernetes/cloud-provider/issues/59

#### Special notes for your reviewer:

Only manually tested. See manual test description in comment below.

#### Does this PR introduce a user-facing change?

```release-note
Fixes a bug which could have allowed an improperly annotated LoadBalancer service to become active.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by mdbooth** [comment]
I have now tested this manually and verified that it fixes the problem.

To describe the problem: cloud-provider-openstack [adds an annotation to the Service object](https://github.com/kubernetes/cloud-provider-openstack/blob/21ce6dd984e5599ba3c411c960acdbcf9b5dbf03/pkg/openstack/loadbalancer.go#L1907) during reconciliation. In our environment we were incorrectly deploying cloud-provider-openstack without the required permission to patch the service object, which resulted in an error message like:

```
I0422 10:30:04.823353   64011 event.go:294] "Event occurred" object="lb-test-ns/lb-test-svc2" fieldPath="" kind="Service" apiVersion="v1" type="Warning" reason="SyncLoadBalancerFailed" message="Error syncing load balancer: failed to ensure load balancer: failed to patch service object lb-test-ns/lb-test-svc2: services \"lb-test-svc2\" is forbidden: User \"system:serviceaccount:kube-system:cloud-controller-manager\" cannot patch resource \"services\" in API group \"\" in the namespace \"lb-test-ns\
```

However, what we observed was that the next time the Service was reconciled after the failure, the reconcile would succeed but the annotation had not been added to the Service object.

The reason for this was that the Service object was dirty from the previous invocation. As the annotation was still present from the previous run there was no difference for the Patcher to patch, and therefore the reconcile erroneously succeeds.

I have verified that with:

* cloud-provider-openstack deployed without the required permissions to patch Service objects
* this patch in place

cloud-provider-openstack will now fail with the above permission error on *every* reconcile attempt, not just the first one.

**Comment by mdbooth** [comment]
I would appreciate advice on how to write an automated test for this. I think the ideal test would be:

* Have the informer return a Service object
* Modify the object in a reconciler, but don't push the modification to apiserver and return an error
* sync the Service a second time, retrieving the object from the informer
* assert that the object is not dirty

It might be possible to construct a unit test to do this but it would be quite complicated and probably not very easy to understand or maintain. It feels more like envtest territory?

Alternatively, we could argue that this would really be a test of the behaviour of informers, not the service controller, and doesn't require a separate test?

**Comment by mdbooth** [comment]
controller-runtime does this here, btw: https://github.com/kubernetes-sigs/controller-runtime/blob/c162794a9b12dea31c076820938a779873e93957/pkg/cache/internal/cache_reader.go#L84-L90

**Comment by lobziik** [comment]
lgtm :)

**Comment by mdbooth** [comment]
@andrewsykim the impact of this that I've noticed is fairly minor and by fixing the underlying permission issue we're unlikely to see it often, although the problem still exists if there's a transient failure to patch the Service object.

I haven't been following the 1.24 release schedule, so I don't know how frozen we are right now. Is it worth trying to merge this, or wait until after 1.24?

**Comment by mdbooth** [comment]
@andrewsykim This is still a thing, btw. Any idea who can approve?

**Review by thockin — COMMENTED** [review]
Thanks!

/lgtm
/approve
